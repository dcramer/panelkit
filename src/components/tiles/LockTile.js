import PropTypes from "prop-types";

import Tile, { TileProps } from "../Tile";

export default class LockTile extends Tile {
  static propTypes = {
    ...TileProps,
    entityId: PropTypes.string.isRequired,
  };

  getWatchedEntityIds() {
    return [this.props.entityId];
  }

  onClick = async () => {
    const { state } = this.getEntity(this.props.entityId);
    await this.callService("lock", state === "locked" ? "unlock" : "lock", {
      entity_id: this.props.entityId,
    });
  };

  getDefaultIcon() {
    const { state } = this.getEntity(this.props.entityId);
    return state === "locked" ? "lock" : "lock-open";
  }
}
