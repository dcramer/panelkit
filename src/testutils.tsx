import React from "react";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { Entity } from "./types";
import HomeAssistant, { SuggestedChanges } from "./hass";

export const URL = "http://localhost:8123";

export const ACCESS_TOKEN = "example.access.token";

export const DEFAULT_CONFIG = {
  url: URL,
  accessToken: ACCESS_TOKEN,
  tiles: [],
};

export const simulateTouch = (container: any) => {
  userEvent.click(container);
};

export const getTileProps = (entity: Entity) => {
  const hass = new HomeAssistant(DEFAULT_CONFIG);

  hass.callService = jest
    .fn()
    .mockImplementation(
      (
        domain: string,
        service: string,
        serviceData: any,
        suggestedChanges: SuggestedChanges | null = null
      ) => {
        return Promise.resolve();
      }
    );
  hass.sendMessage = jest.fn().mockImplementation(() => {
    return Promise.resolve();
  });

  hass.populateEntityCache([entity]);

  return {
    entityId: entity.entity_id,
    hass,
    cameraList: [],
  };
};

export const renderTile = (
  Component: any,
  entity: Entity,
  props: { [key: string]: any } = {}
) => {
  const tileProps = {
    ...getTileProps(entity),
    ...props,
  };

  const result = render(<Component {...tileProps} />);

  return {
    ...result,
    hass: tileProps.hass,
    container: result.container.firstChild,
  };
};
