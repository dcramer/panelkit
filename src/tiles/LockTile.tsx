import Tile, { TileProps } from "./Tile";

type LockTileProps = TileProps & {
  entityId: string;
};

export default class LockTile extends Tile<LockTileProps> {
  onTouch = () => {
    const { state } = this.getEntity(this.props.entityId);
    this.callService(
      "lock",
      state === "locked" ? "unlock" : "lock",
      {
        entity_id: this.props.entityId,
      },
      {
        [this.props.entityId]: {
          state: state === "locked" ? "unlocked" : "locked",
        },
      }
    );
  };

  getDefaultIcon() {
    const { state } = this.getEntity(this.props.entityId);
    return state === "locked" ? "lock" : "lock-open";
  }
}
