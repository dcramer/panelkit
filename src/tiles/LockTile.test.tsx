import { waitFor } from "@testing-library/react";

import { renderTile, simulateTouch } from "../testutils";
import LockTile from "./LockTile";

test("renders", async () => {
  const { container, getByText } = renderTile(LockTile, {
    entity_id: "lock.test",
    state: "unlocked",
    attributes: {
      friendly_name: "Test Lock",
    },
  });

  expect(getByText(/Test Lock/i)).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});

test("unlocks on touch event", async () => {
  const { container, hass } = renderTile(LockTile, {
    entity_id: "lock.test",
    state: "locked",
    attributes: {
      friendly_name: "Test Lock",
    },
  });

  simulateTouch(container);

  await waitFor(() => {
    expect(hass.callService).toHaveBeenCalledTimes(1);
    expect(hass.callService).toHaveBeenCalledWith(
      "lock",
      "unlock",
      {
        entity_id: "lock.test",
      },
      {
        "lock.test": { state: "unlocked" },
      }
    );
  });
});

test("locks on touch event", async () => {
  const { container, hass } = renderTile(LockTile, {
    entity_id: "lock.test",
    state: "unlocked",
    attributes: {
      friendly_name: "Test Lock",
    },
  });

  simulateTouch(container);

  await waitFor(() => {
    expect(hass.callService).toHaveBeenCalledTimes(1);
    expect(hass.callService).toHaveBeenCalledWith(
      "lock",
      "lock",
      {
        entity_id: "lock.test",
      },
      {
        "lock.test": { state: "locked" },
      }
    );
  });
});
