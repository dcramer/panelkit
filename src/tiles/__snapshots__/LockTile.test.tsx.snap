// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div
  class="tile state-unlocked touchable"
  style="cursor: pointer;"
>
  <div
    class="tile-container"
  >
    <div
      class="tile-status"
    >
      Unlocked
    </div>
    <div
      class="tile-body"
    >
      <span
        class="tile-icon"
      >
        <svg />
      </span>
    </div>
    <div
      class="tile-title"
    >
      Test Lock
    </div>
  </div>
</div>
`;
